From 66647e153353bb832ee0a97eefdd7c69f50e2a0e Mon Sep 17 00:00:00 2001
From: Amit Shah <amit.shah@redhat.com>
Date: Tue, 12 Aug 2014 11:30:53 -0400
Subject: [char] hwrng: add randomness to system from rng sources

Message-id: <77dd90b9ac31e53488e6001083058183425d7c91.1407842211.git.amit.shah@redhat.com>
Patchwork-id: 87650
O-Subject: [RHEL7.1 kernel PATCH 3/9] hwrng: add randomness to system from rng sources
Bugzilla: 1119299
RH-Acked-by: Amos Kong <akong@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Herbert Xu <herbert.xu@redhat.com>

From: Kees Cook <keescook@chromium.org>

When bringing a new RNG source online, it seems like it would make sense
to use some of its bytes to make the system entropy pool more random,
as done with all sorts of other devices that contain per-device or
per-boot differences.

Signed-off-by: Kees Cook <keescook@chromium.org>
Reviewed-by: Jason Cooper <jason@lakedaemon.net>
Signed-off-by: Herbert Xu <herbert@gondor.apana.org.au>
(cherry picked from commit d9e79726193346569af7953369a638ee2275ade5)
Signed-off-by: Amit Shah <amit.shah@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/char/hw_random/core.c b/drivers/char/hw_random/core.c
index cf49f1c..7ac96fb 100644
--- a/drivers/char/hw_random/core.c
+++ b/drivers/char/hw_random/core.c
@@ -41,6 +41,7 @@
 #include <linux/miscdevice.h>
 #include <linux/delay.h>
 #include <linux/slab.h>
+#include <linux/random.h>
 #include <asm/uaccess.h>
 
 
@@ -304,6 +305,8 @@ int hwrng_register(struct hwrng *rng)
 {
  int err = -EINVAL;
  struct hwrng *old_rng, *tmp;
+ unsigned char bytes[16];
+ int bytes_read;
 
  if (rng->name == NULL ||
      (rng->data_read == NULL && rng->read == NULL))
@@ -344,6 +347,10 @@ int hwrng_register(struct hwrng *rng)
  }
  INIT_LIST_HEAD(&rng->list);
  list_add_tail(&rng->list, &rng_list);
+
+ bytes_read = rng_get_data(rng, bytes, sizeof(bytes), 1);
+ if (bytes_read > 0)
+  add_device_randomness(bytes, bytes_read);
 out_unlock:
  mutex_unlock(&rng_mutex);
 out:
-- 
1.7.1